### PR TITLE
describe how the tools work and how they fit together

### PR DESCRIPTION
**Pull request by smason**
target audience is mostly developers / those wanting to understand the code base

probably want some more detailed user facing documentation somewhere as well

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+1. CAD-CI, i.e. automatic validation of CAD models. Currently this is
+   just making sure that models are self consistent, no significant
+   overlaps.
+2. Neutronics modelling, i.e. preprocessing CAD models so they can be
+   used with tools like DAGMC.
+
+# Terminology
+
+ * A **solid**s is a 3D volume described by a set of faces, edges, and
```
Comment (johnnonweiler) [inline ARCHITECTURE.md]:
```suggestion
 * A **solid** is a 3D volume described by a set of faces, edges, and
```

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+   verticies.
+ * Solids **intersect** when some point exists that is within either
+   solids or on the surface of either solid. It's useful to
+   distinguish between these cases:
```
Comment (johnnonweiler) [inline ARCHITECTURE.md]:
```suggestion
 * Solids **intersect** when some point exists that is within (or on the surface of) both solids. It's useful to
   distinguish between these cases:
```

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+     empty. For example, a vertex of one solid lies on the face of
+     another. Or, minimally, when two vertices are at the same
+     location.
+   * Intersection solids **overlap** when the volume of the
+     intersection is not empty. E.g. points exist that are within the
+     shapes.
+ * Solids are **distinct** when they do not intersect in any way. For
+   example, their bounding boxes do not overlap. Note that this can
+   get awkward to determine, e.g. a cylinder could have a torus around
+   it whose "hole" is larger diameter than the cylinder.
```
Comment (johnnonweiler) [inline ARCHITECTURE.md]:
```suggestion
   it whose "hole" has a larger diameter than the cylinder.
```

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+The parts of the CAD we really care about are the solid parts and
+their physical properties, like material. Some operations mutate these
+parts (e.g. the imprint step) and we still want to be able to
+reference the parts consistently.
+
+CAD models tend to be hierarchical (it makes it easier to work on them
+like this) but this makes referring to parts awkward. The way I've
+chosen is to "linearise" the hierarchy of solids into a list, and
+maintain this linear order across related files and operations.
```
Comment (johnnonweiler) [inline ARCHITECTURE.md:47]: A key point here is that you're not creating any new solids (e.g. with solid1, solid2, and intersection12), but are adding the intersection to one of the original solids.

Comment (johnnonweiler) [reply ARCHITECTURE.md:47]: Another question is whether you can get the solids out of the STEP file in a deterministic order?

File: ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+allows an existing BREP file to be worked with.
+
+`step_to_brep` contains code to validate that the input file has been
+interpreted correctly, but this might be worth splitting out into a
+separate tool as more input formats are supported.
+
+# Overlap checking
+
+Once the solids have been linearised, it's a simple matter of
+performing $N(N-1)$ checks to determine which solids intersect with
```
Comment (johnnonweiler) [inline ARCHITECTURE.md]: What does the $N(N-1)$ syntax represent?  (It looks like some kind of Markdown formatting, but GitHub doesn't seem to interpret it as that.)

File: ARCHITECTURE.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+The `overlap_checker` tool outputs all intersecting solids to a CSV
+file. The `overlap_collecter` reads this output to create a BREP file
+containing just the overlapping volumes, which can be loaded along
+with the original model to determine where they are.
+
+# Model clean-up
+
+For a model to be used in (e.g.) neutronics modelling, we need to
+remove any overlaps between solids (`imprint_solids`) and merge
+(`merge_solids`) vertexes/edges/faces where appropriate.
```
Comment (johnnonweiler) [inline ARCHITECTURE.md]:
```suggestion
remove any overlaps between solids (`imprint_solids`) and merge
vertexes/edges/faces (`merge_solids`) where appropriate.
```

File: ARCHITECTURE.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+solid or be empty space. Any points within an intersection cause
+ambiguity which cause models to fail. Imprinting aims to automatically
+fix geometries containing overlaps.
+
+Models also tend to assume that transitions between solids occur via
+shared faces. For example, two adjacent cubes should share adjacent
+faces to indicate they touch each other. Otherwise the model might
+interpret this as needing a particle to transition from solid, into
+air, then back into a solid, or maybe that the two solids can slide
+past each other.
```
Comment (johnnonweiler) [inline ARCHITECTURE.md]: I don't understand the bit about two solids sliding past each other.

**Review by johnnonweiler — COMMENTED** [review]
This seems very clear and helpful to me.

File: ARCHITECTURE.md (hunk 4 above)
Comment (smason) [reply ARCHITECTURE.md:47]: getting files out of a STEP file in deterministic order shouldn't matter, that's just done once when converting to BREP.  it's whether running these different utilities iterate over the BREP solids they get them back in the same order

File: ARCHITECTURE.md (hunk 5 above)
Comment (smason) [reply ARCHITECTURE.md]: oops, forgot about that.  Jupyter recognizes it in their notebooks as a latex style equation, forgot that github doesn't support it!

File: ARCHITECTURE.md (hunk 4 above)
Comment (smason) [reply ARCHITECTURE.md:47]: > A key point here

this was what the following paragraph (i.e. "For example, when checking") is giving examples of?

**Pull request merged**